lib/spack/docs/environments.rst

```diff
@@ -0,0 +1,317 @@
+   Consider the following sequence:
+
+   .. code-block:: console
+
+      $ spack install mypackage
+      $ spack load mypackage
+
+   Sometime later, somebody else installs an earlier version of ``mypackage``:
+
+   .. code-block:: console
+
+      $ spack install mypackage@0.9
+
+   Now the ``spack load mypackage`` command will fail because more
+   than one version of ``mypackage`` is installed.  Thus, ``spack
+   load`` is non-deterministic, and not a reliable way to load
+   specific sets of modules one wishes to build and load.
+   Environments solve this problem by creating a localized context
+   that is immune to what other versions of packages might be
+   installed in the global Spack prefix.
+#. Environments separate the steps of (a) choosing what to
+   install, (b) concretizing, and (c) installing.  This allows
+   Environments to remain stable and repeatable, even if Spack packages
+   are upgraded: specs are only re-concretized when the user
+   explicitly asks for it.  It should even be possible to reliably
+   transport environments between different computers running
+   different versions of Spack!
+#. Since re-concretization only occurs at controlled times, the
+   Environment equivalent of many Spack commands are far faster.  For
+   example, ``spack env <env> list -Ilr`` vs. ``spack spec``.
+#. Environments allow several specs to be built at once; a more robust
+   solution than ad-hoc scripts making multiple calls to ``spack
+   install``.
+#. An Environment that is built as a whole can be loaded as a whole.
+   Spack can generate a script of ``module load`` commands that load
+   the appropraite environment.  And that script will work, without
+   running Spack, as long as the underlying modules remain installed.
+   A fast-running ``spack env <env> load`` command would also be
+   possible.
+
+Other packaging systems also provide environments that are similar in
+some ways to Spack environments; for example, `Conda environments
+<https://conda.io/docs/user-guide/tasks/manage-environments.html>`_ or
+`Python Virtual Environments
+<https://docs.python.org/3/tutorial/venv.html>`_.  Spack environments
+some distinctive features:
+
+#. A spec installed "in" an environment is no different form the same
+   spec installed anywhere else in Spack.  Environments are assembled
+   simply by collecting together a set of specs.  Environments are
+   loaded by generating a set of ``module load`` commands.  (In the
+   future, environments can also be assembled into a single tree using
+   techniques from ``spack view``.)
+#. Spack Environments may contain more than one version of the same
+   package; but only a single module for a package may be loaded.
+   Modules that occur in earlier specs listed in an environment take
+   precedence over modules that occur later.
+
+Using Environments
+------------------
+
+Here we follow a typical use case of creating, concretizing,
+installing and loading an environment.
+
+.. note::
```

Comment:
This doesn't seem to be rendering correctly on GitHub.  I'm not familiar with restructured text, so I don't know if that is just GitHub's renderer or if the newlines need to be removed.